engine/cdrs.go

```diff
@@ -513,6 +516,18 @@ func (cdrS *CDRServer) processEvents(evs []*utils.CGREvent,
 	}
 	if refund {
 		for i, cdr := range cdrs {
+			if cdr.CostDetails == nil { // if CostDetails is not populated, look for it inside the previously stored cdr
+				var prevCDRs []*CDR // only one should be returned
+				if prevCDRs, _, err = cdrS.cdrDb.GetCDRs(
+					&utils.CDRsFilter{CGRIDs: []string{cdr.CGRID},
+						RunIDs: []string{cdr.RunID}}, false); err != nil {
+					utils.Logger.Info(
+						fmt.Sprintf("<%s> could not retrieve previously stored CDR, error: <%s>",
+							utils.CDRs, err.Error()))
+					continue
```

Comment:
Here we should return PartiallyExecuted and write error on log since we cannot refund if we cannot find previous CostDetails.